src/sentry/silo/patches/silo_aware_transaction_patch.py

```diff
@@ -46,13 +48,52 @@ def siloed_on_commit(func: Callable[..., Any], using: Optional[str] = None) -> N
     return _default_on_commit(func, using)
 
 
+def is_in_test_case_body() -> bool:
+    """Determine whether the current execution stack is in a test case body.
+
+    This is a best-effort, potentially brittle implementation that depends on private
+    behavior of the current Pytest implementation. We can't necessarily rely on
+    underscore-prefixed method names being used in a stable way.
+
+    Are you landing here because test cases regressed mysteriously after a Pytest
+    upgrade? Check the list of frames and tweak the condition logic to make this
+    function return false as needed. The case `test_is_in_test_case_body` should
+    ensure that you aren't making `validate_transaction_using_for_silo_mode` too
+    permissive.
+
+    An attempt was also made using Pytest fixtures. We can add state changes around
+    the `django_db_setup` fixture, but post-test teardown seems to be too tightly
+    coupled to the test run to insert a fixture between them. Adding something to the
+    `tearDown()` override in Sentry's BaseTestCase may have worked, but would not
+    have helped with standalone test functions. A better solution may nonetheless
+    exist; refactoring is encouraged if you find one.
+
+    This should not be used as a general-purpose utility function. Avoid calling it
+    in places other than `validate_transaction_using_for_silo_mode` if at all possible.
+    """
+    frames = [str(frame) for (frame, _) in traceback.walk_stack(None)]
+
+    def seek(module_path: str, function_name: str) -> bool:
+        """Check whether the named function has been called in the current stack."""
+        pattern = re.compile(rf"/{module_path}\b.*\b{function_name}>$")
+        return any(pattern.search(frame) for frame in frames)
+
+    return seek("_pytest/runner.py", "pytest_runtest_call") and not (
+        seek("django/test/testcases.py", "_pre_setup")
+        or seek("django/test/testcases.py", "_post_teardown")
+    )
+
+
 def validate_transaction_using_for_silo_mode(using: Optional[str]):
     from sentry.models.outbox import ControlOutbox, RegionOutbox
     from sentry.silo import SiloMode
 
     if using is None:
         raise TransactionMissingDBException("'using' must be specified when creating a transaction")
 
+    if not is_in_test_case_body():
```

Comment:
Nit: Explain at the call site here why this assumption is ok, but why it's not ok in other cases.  Basically, just some more context here would help pairs of 👁️

Reply:
👍 Done.